domain/secretbackend/state/state.go

```diff
@@ -404,6 +394,107 @@ WHERE uuid = $M.uuid`, sqlair.M{}, SecretBackendRotationRow{})
 	return domain.CoerceError(err)
 }
 
+// SetModelSecretBackend sets the secret backend for the given model.
+func (s *State) SetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID, backendName string) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	modelBackendUpsertStmt, err := s.Prepare(`
+INSERT INTO model_secret_backend
+	(model_uuid, secret_backend_uuid)
+VALUES (
+	$M.model_uuid,
+	(SELECT uuid FROM secret_backend WHERE name = $M.backend_name)
+)
+ON CONFLICT (model_uuid) DO UPDATE SET
+	secret_backend_uuid = EXCLUDED.secret_backend_uuid`, sqlair.M{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err = tx.Query(ctx,
+			modelBackendUpsertStmt,
+			sqlair.M{"model_uuid": modelUUID, "backend_name": backendName},
+		).Run()
+		if database.IsErrConstraintForeignKey(err) {
+			return fmt.Errorf("%w: either model %q or secret backend %q", errors.NotFound, modelUUID, backendName)
+		}
+		if err != nil {
+			return fmt.Errorf("setting secret backend %q for model %q: %w", backendName, modelUUID, err)
+		}
+		return nil
+	})
+	return domain.CoerceError(err)
+}
+
+// GetModelSecretBackend returns the configured secret backend for the given model.
+func (s *State) GetModelSecretBackend(ctx context.Context, modelUUID coremodel.UUID) (string, error) {
+	m, err := s.GetModel(ctx, modelUUID)
+	return m.SecretBackendID, errors.Trace(err)
+}
+
+// GetCloudCredential returns the cloud credential for the given model.
+func (s *State) GetCloudCredential(ctx context.Context, modelUUID coremodel.UUID) (cloud.Cloud, cloud.Credential, error) {
+	var (
+		cld  cloud.Cloud
+		cred cloud.Credential
+	)
+
+	db, err := s.DB()
+	if err != nil {
+		return cld, cred, errors.Trace(err)
+	}
+	modelStmt, err := s.Prepare(`
+SELECT &ModelCloudCredentialRow.*
+FROM v_model
+WHERE uuid = $M.uuid`, sqlair.M{}, ModelCloudCredentialRow{})
+	if err != nil {
+		return cld, cred, errors.Trace(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
```

Comment:
@ycliuhw @wallyworld This is a giant potato that runs this SQL:
(from `v_model`)
```sql
SELECT m.uuid,
       mm.cloud_uuid,
       c.uuid        AS cloud_uuid,
       c.name        AS cloud_name,
       cr.uuid       AS cloud_region_uuid,
       cr.name       AS cloud_region_name,
       cc.uuid       AS cloud_credential_uuid,
       cc.name       AS cloud_credential_name,
       cc.owner_uuid AS cloud_credential_owner_uuid,
       cco.name      AS cloud_credential_owner_name,
       ccn.name      AS cloud_credential_cloud_name,
       mm.model_type_id,
       mt.type       AS model_type_type,
       mm.name,
       mm.owner_uuid,
       u.name        AS owner_name,
       l.value       AS life
FROM model_list m
INNER JOIN model_metadata mm ON m.uuid = mm.model_uuid
INNER JOIN cloud c ON mm.cloud_uuid = c.uuid
LEFT JOIN cloud_region cr ON mm.cloud_region_uuid = cr.uuid
LEFT JOIN cloud_credential cc ON mm.cloud_credential_uuid = cc.uuid
INNER JOIN user cco ON cc.owner_uuid = cco.uuid
LEFT JOIN cloud ccn ON cc.cloud_uuid = ccn.uuid
INNER JOIN model_type mt ON mm.model_type_id = mt.id
INNER JOIN user u ON mm.owner_uuid = u.uuid
INNER JOIN life l ON mm.life_id = l.id;
```
(from `LoadClouds`)
```sql
WITH controllers AS (SELECT model_metadata.cloud_uuid
        FROM model_metadata
		INNER JOIN user
        ON user.uuid = model_metadata.owner_uuid
        AND model_metadata.name = $M.controller_name
        AND user.name = $M.controller_user_name)
SELECT (cloud.uuid, cloud.name, cloud_type_id, 
       cloud.endpoint, cloud.identity_endpoint, 
       cloud.storage_endpoint, skip_tls_verify) AS (&Cloud.*),
       IIF(controllers.cloud_uuid IS NULL, false, true) AS &Cloud.is_controller_cloud,
       auth_type.type AS &M.cloud_auth_type, cloud_type.type AS &M.cloud_type
FROM   cloud
       LEFT JOIN cloud_auth_type 
            ON cloud.uuid = cloud_auth_type.cloud_uuid
       JOIN auth_type 
            ON auth_type.id = cloud_auth_type.auth_type_id
       JOIN cloud_type 
            ON cloud_type.id = cloud.cloud_type_id
       LEFT JOIN controllers
            ON controllers.cloud_uuid = cloud.uuid
```
(from `LoadCloudCredentials`)
```sql
SELECT (cc.uuid, cc.name,
       cc.revoked, cc.invalid, 
       cc.invalid_reason, 
       cc.owner_uuid) AS (&Credential.*),
       auth_type.type AS &AuthType.*,
       cloud.name AS &Cloud.*,
       (cc_attr.key, cc_attr.value) AS (&credentialAttribute.*)
FROM   cloud_credential cc
       JOIN auth_type ON cc.auth_type_id = auth_type.id
       JOIN cloud ON cc.cloud_uuid = cloud.uuid
	   JOIN user on cc.owner_uuid = user.uuid
       LEFT JOIN cloud_credential_attributes cc_attr ON cc_attr.cloud_credential_uuid = cc.uuid
```
It uses the same tables repeatedly and needlessly materialises results that are unused.
It can be replaced with a single query along the lines of:
```sql
SELECT * -- What you need.
FROM   model_metadata m
       JOIN cloud c ON m.cloud_uuid = c.uuid
       JOIN cloud_type ct ON c.cloud_type_id = ct.id
       -- From LoadClouds, but why LEFT? Can a cloud not have a cloud auth type?
       LEFT JOIN cloud_auth_type cat ON c.uuid = cat.cloud_uuid
       JOIN auth_type at ON cat.auth_type_id = at.id
       -- I think this will work as the IIF used in LoadClouds, without needing the inline view.
       -- Do you even need the join? Isn't the model name enough to tell you?
       LEFT JOIN user cu ON m.owner_uuid = cu.uuid
            AND cm.name = 'controller'
            AND cu.name = 'admin'
       JOIN cloud_credential cc ON m.cloud_credential_uuid = cc.uuid
WHERE  m.model_uuid = ?
```

This goes to my point about the needs of a domain being served by SQL from that domain. We don't need to be sharing these queries cross-domain, because domains should be acquiring the same info if they need it, in different queries.

Please do a follow up to this _and_ review the SQL in `LoadClouds`.